Use subnetIdSelector.matchLabels for subnet selection in EKS composition

MichaelVL opened the pull request:
### What does this PR do?

This PR allows the composition in `compositions/aws-provider/eks` to use labels for automatically selecting subnets and replaces the current approach where subnet IDs manually needs to be inserted into the list of subnets. See also issue #32 

### Motivation

The EKS cluster composition in `compositions/aws-provider/eks` require manual specification of subnet Ids. In Kubernetes, labels are typically used to reference resources, i.e. labelling subnets and using the `resourcesVpcConfig.subnetIdSelector.matchLabels` feature of the `cluster.eks.crossplane.io` resource would be more convenient and less error prone. Prior art for this can be found in https://github.com/upbound/platform-ref-aws

The implementation in this PR implements the following

- Adds a new parameter `networkId` to the `xvpcsubnets` composition, which is propagated to the subnet label `network.awsblueprints.io/network-id`. The idea behind NOT using the VPC name is that this is a logical network name and may consist of more than a VPC. The label use dash-case as is common for Kubernetes labels.
- Similarly adds a `networkId` to the `xamazoneks` composition and which is used for subnet selection in the cluster `resourcesVpcConfig` settings and similarly for the node group.
- Updates example claims with the `networkId` parameter.

### More

- [X ] Yes, I have tested the PR using my local account setup  (Provide any test evidence report under Additional Notes)

- [X ] Yes, I have added a new example under [examples](https://github.com/aws-samples/crossplane-aws-blueprints/tree/main/examples) to support my PR  (partly, since this PR updates existing examples)

- [ ] Yes, I have updated the [docs](https://github.com/aws-samples/crossplane-aws-blueprints/tree/main/doc) for this feature

- [X ] Yes, I have linked to an issue or feature request (applicable to PRs that solves a bug or a feature request)

**Note**:
 - Not all the PRs require examples and docs
 - We prefer small, well tested pull requests. Please ensure your pull requests are self-contained, and commits are squashed

### For Moderators
- [ ] E2E Test successfully complete before merge?

### Additional Notes

This implementation was inspired by https://github.com/upbound/platform-ref-aws

nabuskey commented:
Hey @MichaelVL. Thank you so much for this PR. I agree we should have an example of composition which uses subnet selector, but I think it's a good idea to have both examples available instead of just one. It is possible that networking setup and EKS cluster management may be handled differently. E.g. Networking may require more complex setup such as Transit Gateway with manual on-premises network connections. 
So I think it makes sense to have two compositions which satisfy the same XRD.

MichaelVL commented:
Hi. I agree, that both approaches for specifying networks will be highly useful. My only concern is, that the two compositions are not 100% identical, i.e. one will use the `subnetIDs` parameter, the other the `networkId` parameter. So they do not exactly comply to the same schema, annd we would have to add both parameters to the XRD and default these parameters such that they can be left unspecified for the compositions that do not use the specific parameter. This would again weaken error checking.

Ideally they should be two separate compositions with exact schemas. The draw-back with that would be a lot of duplication.

FYI, I have added an implementation of the 'two compositions satisfying one XRD' here: https://github.com/MichaelVL/crossplane-aws-blueprints/tree/feature/subnet-label-selector-multicomposition

MichaelVL commented:
I updated the PR with the additional compositions. Unfortunately, the patch `policy` feature only works in the `CombineFromCompostite` patch type, so 'missing parameter' checking is not possible.

I also restored the original vpc-subnet and eks examples (the one using subnetIds) and only updated them with labels to select the original compositions. I added a new example that use the networkId label-based subnet selection.

nabuskey reviewed (approved):
Thank you so much for this PR! Changes look good to me.
